pkg/proxy/proxy.go

```diff
@@ -19,28 +19,89 @@ import (
 	"github.com/redhat-appstudio/sprayproxy/pkg/metrics"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
+	"golang.org/x/exp/slices"
 )
 
-type BackendsFunc func() []string
+type backend struct {
+	URL string `json:"url"`
+}
 
 type SprayProxy struct {
-	backends    BackendsFunc
-	insecureTLS bool
-	logger      *zap.Logger
+	backends              []string
+	insecureTLS           bool
+	enableDynamicBackends bool
+	logger                *zap.Logger
 }
 
-func NewSprayProxy(insecureTLS bool, logger *zap.Logger, backends ...string) (*SprayProxy, error) {
-	backendFn := func() []string {
-		return backends
-	}
-
+func NewSprayProxy(insecureTLS bool, enableDynamicBackends bool, logger *zap.Logger, backends ...string) (*SprayProxy, error) {
 	return &SprayProxy{
-		backends:    backendFn,
-		insecureTLS: insecureTLS,
-		logger:      logger,
+		backends:              backends,
+		insecureTLS:           insecureTLS,
+		enableDynamicBackends: enableDynamicBackends,
+		logger:                logger,
 	}, nil
 }
 
+func (p *SprayProxy) RegisterBackends(c *gin.Context) {
+	zapCommonFields := []zapcore.Field{
+		zap.String("method", c.Request.Method),
+		zap.String("path", c.Request.URL.Path),
+		zap.String("query", c.Request.URL.RawQuery),
+		zap.Bool("dynamic-backends", p.enableDynamicBackends),
+	}
+	if p.enableDynamicBackends {
+		var newUrl backend
+		if err := c.ShouldBindJSON(&newUrl); err != nil {
+			c.String(http.StatusBadRequest, "please provide a valid json body")
+			p.logger.Info("backend server register request to proxy is rejected, invalid json body", zapCommonFields...)
+			return
+		}
+		zapBackendFields := append(zapCommonFields, zap.String("backend", newUrl.URL))
+		if !slices.Contains(p.backends, newUrl.URL) {
+			p.backends = append(p.backends, newUrl.URL)
```

Comment:
I think this code is susceptible to race conditions. The `backends` array can be accessed by more than one goroutine in production.

I recommend that we move the management of backend state to a separate package `/pkg/backends`, so that we can tune the storage and retrieval of data.